opencog/query/InitiateSearchCB.cc

```diff
@@ -345,10 +454,39 @@ bool InitiateSearchCB::neighbor_search(PatternMatchEngine *pme)
 	// Note also: the user is allowed to specify patterns that have
 	// no constants in them at all.  In this case, the search is
 	// performed by looping over all links of the given types.
+#if USE_PATTERN_TERM
+	PatternTermPtr start_term = find_thinnest_term(clauses);
+
+	// Cannot find a starting point! This can happen if:
+	// 1) all of the clauses contain nothing but variables,
+	// 2) all of the clauses are evaluatable(!),
+	// Somewhat unusual, but it can happen.  For this, we need
+	// some other, alternative search strategy.
+	if (start_term == PatternTerm::UNDEFINED) {
+		_search_fail = true;
+		return false;
+	}
+	// If only a single choice, fake it for the loop below.
+	if (0 == _choices.size())
+	{
+		Choice ch;
+		ch.thinnest_term = start_term;
+		_choices.push_back(ch);
+	}
+#else
 	size_t bestclause;
 	Handle best_start = find_thinnest(clauses, _pattern->evaluatable_holders,
 	                                  _starter_term, bestclause);
+#endif
+#if USE_PATTERN_TERM
 
+	for (const Choice& choice : _choices)
+	{
+		bool found = pme->explore_term(choice.thinnest_term);
+		if (found)
+			return true;
+	}
+#else
```

Comment:
That's a hell of a simplification. I suppose it also means you're on your way to enable parallelization for backtracking, isn't it?

Reply:
Not sure I understand you point correctly. Do you mean that this code is oversimplified comparing to previous version? I am going to revisit it when approach is confirmed and unit tests are green as I would like make sure this approach works first.

May be you mean that it was a place where matching was parallelized before? Now only place to explore incoming set is PatternMatchEngine::explore_up_branches. I didn't think what does it mean for parallelizing but PatternMatchEngine::explore_up_branches can be parallelized as well.

Reply:
Oh, I definitely meant "hell" in a good way.

Regarding parallelization, yes that is what I meant, I think it's a good thing and it forces us to address the problem at its core. It makes it clear that the initial suggestion we made to you about parallelization is just a subcase of backtracking parallelization.